pkg/controllers/disruption/consolidation.go

```diff
@@ -192,6 +193,52 @@ func (c *consolidation) computeConsolidation(ctx context.Context, candidates ...
 		results.NewNodeClaims[0].Requirements.Add(scheduling.NewRequirement(v1beta1.CapacityTypeLabelKey, v1.NodeSelectorOpIn, v1beta1.CapacityTypeSpot))
 	}
 
+	results.NewNodeClaims[0].NodeClaimTemplate.InstanceTypeOptions = instanceTypesOptionsFilteredByPrice
+	return Command{
+		candidates:   candidates,
+		replacements: results.NewNodeClaims,
+	}, nil
+}
+
+// Compute command to execute spot-to-spot consolidation if:
+// 1. The SpotToSpotConsolidation feature flag is set to true.
+// 2. There are atleast 15 cheapest instance type replacement options to consolidate.
+// 3. The current candidate is NOT part of the first 15 cheapest instance types inorder to avoid repeated consolidation for single-node consolidation.
+func (c *consolidation) computeSpotToSpotConsolidation(ctx context.Context, candidates []*Candidate,
+	results *pscheduling.Results, instanceTypesOptionsFilteredByPrice []*cloudprovider.InstanceType) (Command, error) {
+
+	if !options.FromContext(ctx).FeatureGates.SpotToSpotConsolidation {
+		if len(candidates) == 1 {
+			// Spot consolidation is turned off.
+			c.recorder.Publish(disruptionevents.Unconsolidatable(candidates[0].Node, candidates[0].NodeClaim, "SpotToSpotConsolidation is disabled, can't replace a spot node with a spot node")...)
+		}
+		return Command{}, nil
+	}
+
+	// Allow single-node/multi-node consolidation only if there are minimum number of possible InstanceTypes for a NodeClaim
+	if len(results.NewNodeClaims[0].NodeClaimTemplate.InstanceTypeOptions) < MinInstanceTypesForSpotToSpotConsolidation {
+		// InstanceType flexibility is not good enough to replace spot node.
+		c.recorder.Publish(disruptionevents.Unconsolidatable(candidates[0].Node, candidates[0].NodeClaim, fmt.Sprintf("SpotToSpotConsolidation requires %d instance type replacement options to consolidate",
+			MinInstanceTypesForSpotToSpotConsolidation))...)
+		return Command{}, nil
+	}
+
+	// All possible replacements for the current candidate sorted by price
+	instanceTypesOrderedByPrice := results.NewNodeClaims[0].NodeClaimTemplate.InstanceTypeOptions.OrderByPrice(results.NewNodeClaims[0].Requirements)
```

Comment:
This pricing ordering needs to take into account only the instance types that have spot offerings, right? So, I think you need to filter down the options to the ones that just have spot and then order those by their price. Otherwise, you get into a state where you could have 15 instance types, but only half of them have spot offerings for the flexibility requirement

Reply:
I see. Good find, Jonathan. While implementing this, for some reason, I had an impression that replacement offerings simulated for a spot candidate would only contain spot offering but if it has both spot, od based on the nodepool requirement, then yeah we will have to filter down the options to the ones that just have spot and then order those by price. Otherwise, it will be a problem like you stated.

Reply:
Addressed this.